activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -534,6 +532,15 @@ def insert_fixtures_set(fixture_set, tables_to_delete = [])
         end
       end
 
+      protected
+
+        def check_version
```

Comment:
Why `protected`? Is this `check_version` is called from another connection?
Isn't enough `private`?

Reply:
Yeah, the use of `protected` is to signal to future subclasses that it's possible to modify but not available within the public API. This prevents other objects from using and restricts behavior to subclasses only. I also want to be able to subclass and override the `check_version` method in this superclass so the check can be done at a later time. This change allows that to happen.

Reply:
It still sounds like `private` is enough.
It is possible to override and use the `check_version` even if `private`, and I can't find there is any use case that a connection call another connection's `check_version` at this point.

Reply:
That is true, Ruby would let us do that, unfortunately. 😅 The problem is that, in this situation, the use of `private` would break encapsulation of the superclass. In OO design, `private` is meant to encapsulate the inner workings of a class and let readers of the code know these private methods are meant for the superclass alone. By using `protected`, I'm signaling to future developers it's OK to subclass and modify as long as they don't violate the [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) in S.O.L.I.D. design (i.e. the L in S.O.L.I.D.) That was long winded but hopefully explains better my reasoning and considerations to the design.

Reply:
In Ruby, private and protected have a well-defined meaning, and it's not that.

Reply:
BTW, there is currently no subclasses that need `check_version`...yet. That's what I'm working on (downstream from Rails). :wink:

Reply:
In Ruby, `protected` is designed to expose a method to another instance.
We do not use it for any other purpose.

Reply:
:bow: @kamipo, @matthewd I was not aware of this practice, good to know. BTW, might be worthwhile to mention this in the [Rails API Guidelines](https://guides.rubyonrails.org/api_documentation_guidelines.html#method-visibility) so others don't stumble across this?

Reply:
It's the Rails API *Documentation* Guidelines.
I'm not excited to mention about Ruby's semantics in Rails guides.
Please see the docs Ruby's itself instead.

http://docs.ruby-doc.com/docs/ProgrammingRuby/html/tut_classes.html#S4

Reply:
:bow: @kamipo